### ISSUE TITLE
No proxy support: context deadline exceeded

### ISSUE THREAD

**Issue by revolveyao** [PaulSonOfLars/gotgbot#16]
# Description of the problem / feature request

run `gotgbot.NewBot` panic : failed to execute GET request to getMe: context deadline exceeded

# System information:

- Operating System + version: windows 10
- Output of `go version`: 1.16.2
- Library version (as seen in `go.mod` file): v2.0.0-beta6

# Any other information, logs, or outputs that you would like to share?

If you have large files to share, please upload as attachment, or provide a (public) link.

**Comment by PaulSonOfLars** [comment]
Hey, thanks for raising an issue!

The getMe request timing out is usually caused by one of 2 things:
1) poor connection to telegram servers. Try running your bot in a datacenter closer to the bot API.
2) the token hasn't been used in a while. This means that telegram needs to start your bot on their side, and the getMe request may take longer than usual. In this case, retrying a few seconds later should fix the issue.

To address 2), I think it could be a good idea for the library to have an increased the initial timeout for the first getMe request.
What do you think?

**Comment by revolveyao** [comment]
Hey, thanks for reply!

But I didn't have a problem using my browser to open the same URL.
I also feel the need to add the initial timeout.

**Comment by PaulSonOfLars** [comment]
Because your browser doesn't have a timeout like the lib does. Timeouts are necessary to avoid the bot "hanging" and waiting for long periods of time due to poor network.

Also, you might have opened the URL after 2), so it would've been faster. Retrying to run the script should also work, after telegrams initial startup delay.

**Comment by revolveyao** [comment]
I almost know the problem, it should be that the program did not go system proxy.

**Comment by PaulSonOfLars** [comment]
I'm not sure I follow what you're saying, could you rephrase?

**Comment by revolveyao** [comment]
My network doesn't have direct access to Telegram's website and requires a proxy. Isn't the golang http client not using a proxy?

**Comment by PaulSonOfLars** [comment]
The library uses the default device network settings.

If you need a proxy, we might need to change the http client that the bot struct uses to support it. I think this would be a great addition, though I don't have a way to test locally.
If I open up a branch with the changes for connecting via proxy, would you be able to test?

**Comment by revolveyao** [comment]
I can help with the test.
Then I think I can expose the default timeout in the NewBot method.

**Comment by PaulSonOfLars** [comment]
If you try the above branch (`go get github.com/PaulSonOfLars/gotgbot/v2@feat/proxy-support`), you should be able to setup a SOCKS5 proxy by editing the transport field in the `http.Client`. See [this gist](https://gist.github.com/ometa/71d23ed48c03c003f6e4910648612859) for more context on how to do that.

That branch also increases the GetMe timeout to be 10 seconds instead of the previous 3; that should also help you.

Let me know if this works :)

**Comment by revolveyao** [comment]
When I use go get the times are wrong :invalid version: version "feat/proxy-support" invalid: disallowed version string.
I can't solve this problem.

**Comment by revolveyao** [comment]
I solved the problem using commit sha1.

**Comment by revolveyao** [comment]
I've tested it out, getMe can call successfully after using the proxy, but calling getUpdates is wrong and should be caused by not using the proxy.

**Comment by PaulSonOfLars** [comment]
Ah, yes 🤦‍♂️ the sha was the one to use, my bad.

But good, glad to hear the getMe worked! Step in the right direction. I know what's wrong with the updater, should be an easy fix. I'll take care of that later.

Thank you for your help!

**Comment by revolveyao** [comment]
Thank you for your work. It's my pleasure to help.

**Comment by PaulSonOfLars** [comment]
Hey @revolveyao - if you could try out the latest commit (`9ef8d0a`) and confirm it works, i'll merge in the PR and cut a new beta release.

And if you could do me a favour, I'd love to have an example of the code you use to setup the proxy so I can document it!

**Comment by revolveyao** [comment]
I experimented, or reported the mistake: `failed to get updates; sleeping 1s: failed to execute GET request to getUpdates: Get "https://api.telegram.org/bot<token>/getUpdates?limit=0&offset=-1&timeout=0": context deadline exceeded`

That's how I added the proxy, is there a problem?
``` go
	dialer, err := proxy.SOCKS5("tcp", "127.0.0.1:10808", nil, proxy.Direct)
	if err != nil {
		fmt.Fprintln(os.Stderr, "can't connect to the proxy:", err)
		os.Exit(1)
	}
	// setup a http client
	httpTransport := &http.Transport{}

	httpClient := http.Client{Transport: httpTransport}
	// set our socks5 as the dialer
	httpTransport.Dial = dialer.Dial
	b, err := gotgbot.NewBot(conf.Conf.Token, &gotgbot.BotOpts{
		Client:      httpClient,
		GetTimeout:  gotgbot.DefaultGetTimeout,
		PostTimeout: gotgbot.DefaultPostTimeout,
	})
```

**Comment by PaulSonOfLars** [comment]
hmm, damn.

I've just pushed a new commit (`1a81a17`) which should fix a small bug to do with timeouts in the call to `startPolling()` when working with slower connections (or proxies).
Does that fix it?

**Comment by revolveyao** [comment]
Yes, it's working now.

**Comment by PaulSonOfLars** [comment]
Awesome! Will merge and cut a new beta release (`v2.0.0-beta7`).

Feel free to join the support chat if you have any feedback or questions regarding the library!
